HPCC-24218 Maintain cross ESDL transform state and add new functions

afishbeck opened the pull request:
Use a scriptContext to support keeping script state for the entire transaction.

Allowing for the upcomming addition of multiple script entry points.

Add new script functions to add to scriptContext state:
    set-log-profile - signal which logging profile should be followed
    set-log-option - control logging options
    store-value - store values that can be accessed by other scripts

Add new script functions (single node operations for now):
   remove-node - remove a single node from the target tree
   rename-node - rename a single node on the target tree

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexisrisk.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [x] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

timothyklemm commented on system/xmllib/libxml_xpathprocessor.cpp:
```diff
@@ -775,6 +789,352 @@ extern ICompiledXpath* compileXpath(const char * xpath)
+    virtual void setAttribute(const char *section, const char *name, const char *value) override
+    {
+        xmlNodePtr sect = ensureSection(section);
+        xmlNewProp(sect, (const xmlChar *)name, (const xmlChar *)value);
```
Not sure this is sufficient. Consider a use case where an attribute could be set to a default, then a subsequent operation intends to override the default. My observation is that the first operation succeeds and the subsequent operation is ignored. xmlSetProp will update a node, but may not (documentation isn't clear) insert.

On a tangent, it might also make sense to have operations that can remove an attribute (using xmlUnsetProp). The ability to remove a previously applied attribute might be the preferred approach to overriding.

afishbeck replied:
Agreed I'll change it to xmlSetProp.  I've looked at the code and that does create or update.

I'll also look at removal.

afishbeck replied:
Changing to xmlSetProp.  And changing this function to remove a prop if value is nullptr.  But leaving adding that capability to scripts for future consideration.

timothyklemm commented on esp/esdllib/esdl_script.cpp:
```diff
@@ -98,9 +101,11 @@ class CEsdlTransformOperationVariable : public CEsdlTransformOperationBase
 public:
     CEsdlTransformOperationVariable(IPropertyTree *tree, const StringBuffer &prefix) : CEsdlTransformOperationBase(tree, prefix)
     {
+        if (m_traceName.isEmpty())
+            m_traceName.set(tree->queryProp("@name"));
```
This seems to be redundant, to the point that it could probably make sense moving in to the base class. If the base class handles m_traceName, it makes sense for it to handle fall-back options (could maybe add an optional fall-back name to the constructor parameter list to not assume).

I'm not sure I would go so far as to move storage for @name into the base, although I probably wouldn't complain if it happened.

afishbeck replied:
I'll take a look but @name is only a fallback for trace on commands where it is already expected.  name is too valuable as a parameter to use it to mean trace.  So I want to avoid that in the future.  When dealing with name + value situations it's a shame if we can't make it intuitive because of how we did tracing.

asselitx commented on esp/esdllib/esdl_script.cpp:
```diff
@@ -769,155 +1046,145 @@ class CEsdlCustomTransform : public CInterfaceOf<IEsdlCustomTransform>
+    //in case transform wants to make use of these values:
+    //make them few well known values variables rather than inputs so they are automatically available
+    StringBuffer temp;
+    xpathContext->addVariable("query", scriptCtx->getXPathString("target/*/@queryname", temp));
```
Why not also move this statement for the 'query' variable into the IXpathContext* createXpathContext() function in libxml_xpathprocessor.cpp where the other similar statements moved?

afishbeck replied:
may move this in the future.  I'm just not sure where I want knowledge of the Target config section.  The other values are all from the ESDL section which comes from specific values in the ESDL definition.

asselitx commented on system/xmllib/libxml_xpathprocessor.cpp:
```diff
@@ -775,6 +789,352 @@ extern ICompiledXpath* compileXpath(const char * xpath)
+            throw MakeStringException(-1, "CEsdlScriptContext:createXpathContext: section not found %s", section);
+        CLibXpathContext *xpathContext = new CLibXpathContext(doc, sect, strictParameterDeclaration);
+        StringBuffer val;
+        xpathContext->addVariable("method", getAttribute("esdl", "method", val));
```
Why did you spit  up the initialization of the xpathContext between this createXpathContext() function and the processServiceAndMethodTransforms() method? I found it nice to see all the context setup in that one function- made it easy to tell which values were treated like parameters and which like implicit variables.

afishbeck replied:
I thought these values could be understood by the script context, but wasn't sure about some others.  We can consider reorganizing further in the future.

asselitx commented on esp/esdllib/esdl_xpath_extensions_libxml.cpp:
```diff
@@ -176,7 +189,163 @@ static void getFeatureSecAccessFlagsFunction (xmlXPathParserContextPtr ctxt, int
+        return;
+    }
+
+    if (nargs != 1)
```
Why must nargs==1? It is not used for anything?

asselitx commented on system/xmllib/libxml_xpathprocessor.cpp:
```diff
@@ -775,6 +789,352 @@ extern ICompiledXpath* compileXpath(const char * xpath)
+void copyXmlNode(IPropertyTree *tree, xmlNodePtr node)
+{
+    for(xmlAttrPtr att=node->properties;att!=nullptr; att=att->next)
+        tree->setProp((const char *)att->name, (const char *)xmlGetProp(node, att->name));
```
Do we need to manage the memory of the xmlChar* returned from xmlGetProp here? Or is it just returning a pointer to the xmlChar* it retains ownership of?

timothyklemm replied:
We own the block returned from xmlGetProp, and are required to free it.

asselitx commented on system/xmllib/libxml_xpathprocessor.cpp:
```diff
@@ -775,6 +789,352 @@ extern ICompiledXpath* compileXpath(const char * xpath)
+        return nullptr;
+    if (att->type != XML_ATTRIBUTE_NODE || att->children==nullptr || (att->children->type != XML_TEXT_NODE && att->children->type != XML_CDATA_SECTION_NODE))
+        return nullptr;
+    return (const char *) att->children->content;
```
So an attribute's value is its child node if that node is TEXT or CDATA?

afishbeck replied:
It's of type XML_ATTRIBUTE_NODE content.  A CDATA node is a different type.  In this "queryAttrValue() function I am relying on the fact that these are simple attributes and not a compounding of multiple parts like some elements and perhaps more complex attributes might be.  If we try and use this function for uses that don't match this assumption we may need to make it less efficient and have to allocate and delete memory.  For now this one method errors on the side of efficiency.

asselitx commented on system/xmllib/libxml_xpathprocessor.cpp:
```diff
@@ -775,6 +789,352 @@ extern ICompiledXpath* compileXpath(const char * xpath)
+class CEsdlScriptContext : public CInterfaceOf<IEsdlScriptContext>
+{
+private:
+    void *espCtx = nullptr;
```
Why not store as an IEspContext *?

afishbeck replied:
Exactly I didn't want to expose the common library to ESP internals.  We could move things around or share some fairly isolated headers.  I think IEspContext may carry a lot of baggage.

asselitx commented on system/xmllib/libxml_xpathprocessor.cpp:
```diff
@@ -775,6 +789,352 @@ extern ICompiledXpath* compileXpath(const char * xpath)
+
+        xmlParserCtxtPtr parserCtx = xmlCreateDocParserCtxt((const unsigned char *)xml);
+        if (!parserCtx)
+            throw MakeStringException(-1, "CEsdlScriptContext:setXmlSection: Unable to init parse of %s XML content", section);
```
Error message refers to incorrect function name.

asselitx commented on system/xmllib/libxml_xpathprocessor.cpp:
```diff
@@ -775,6 +789,352 @@ extern ICompiledXpath* compileXpath(const char * xpath)
+        parserCtx->node = sect;
+        xmlParseDocument(parserCtx);
+        int wellFormed = parserCtx->wellFormed;
+        xmlFreeDoc(parserCtx->myDoc); //dummy document
```
Should this still be in here? I see you commented out a line above setting parserCtx->myDoc.

afishbeck replied:
Setting myDoc above ends up serving no purpose but leaks this dummy doc.  So I had to do it this way not to leak.  If we ever find a problem I will have to switch gears, parse into a new doc, and then adopt the content into the main doc.  But I think this actually works if a bit confusing.

asselitx commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -1447,7 +1447,21 @@ void EsdlServiceImpl::prepareFinalRequest(IEspContext &context,
+        scriptContext->setAttribute("esdl", "service", srvdef.queryName());
+        scriptContext->setAttribute("esdl", "method", mthdef.queryMethodName());
+        scriptContext->setAttribute("esdl", "request_type", mthdef.queryRequestType());
+        scriptContext->setAttribute("esdl", "request", mthdef.queryRequestType());
```
Why do we have different attributes for request and request_type?

afishbeck replied:
We use them as the same right now but I could see them sometimes changing.  We just happen to support the request tag that is named after the typename.  May never need to separate but I wanted to make it clear that they may not always be the same.

asselitx commented on system/xmllib/libxml_xpathprocessor.cpp:
```diff
@@ -775,6 +789,352 @@ extern ICompiledXpath* compileXpath(const char * xpath)
+        addChildFromPtree(sect, *tree, tree->queryName());
+    }
+
+    virtual void setAttribute(const char *section, const char *name, const char *value) override
```
Maybe helpful looking toward the future, perhaps we should use some enumerations or constants for the section and name values used here. We might not want to *limit* it to an enumeration, but some standard constants would help with consistency.

afishbeck replied:
I agree we could use constants but they would probably be in the esdl binding.  The esdl script can handle any names, the binding sets our model for what our entry points actually are.  Although some may be common like "store" so could consider some being in esdl_script and some being in the esdl service/binding.

HPCCSmoketest commented:
Process of PR-13848, label: esdlScriptContextData starts now.
The reason of this test is: Forced to re-test.
Commit ID: 954dbeb4622363ba8678e409c06ff751d95a8bcb
Estimated completion time is ~0.8 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-247.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-247.ca-central-1.compute.internal
Sha: 954dbeb4622363ba8678e409c06ff751d95a8bcb
Build: success
Milestone:Install hpccsystems-platform-community_7.9.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |137 |137 |0 |0 |0 |72 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |29 sec  (00:00:29) |
|setup (thor) |11 |11 |0 |44 sec  (00:00:44) |
|setup (roxie) |11 |11 |0 |19 sec  (00:00:19) |
|test (hthor) |882 |882 |0 |316 sec  (00:05:16) |
|test (thor) |820 |820 |0 |769 sec  (00:12:49) |
|test (roxie) |961 |961 |0 |416 sec  (00:06:56) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  737 sec (00:12:17) |  118 sec (00:01:58) |   23 sec (00:00:23) |   19 sec (00:00:19) | 1850 sec (00:30:50) |   18 sec (00:00:18) | 2774 sec (00:46:14) |

HPCCSmoketest commented:
Process of PR-13848, label: esdlScriptContextData starts now.
The reason of this test is: Forced to re-test.
Commit ID: 954dbeb4622363ba8678e409c06ff751d95a8bcb
Estimated completion time is ~0.8 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-221.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-221.ca-central-1.compute.internal
Sha: 954dbeb4622363ba8678e409c06ff751d95a8bcb
Build: success
Milestone:Install hpccsystems-platform-community_7.9.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |137 |137 |0 |0 |0 |72 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |8 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |25 sec  (00:00:25) |
|setup (thor) |11 |11 |0 |43 sec  (00:00:43) |
|setup (roxie) |11 |11 |0 |19 sec  (00:00:19) |
|test (hthor) |882 |882 |0 |320 sec  (00:05:20) |
|test (thor) |820 |820 |0 |764 sec  (00:12:44) |
|test (roxie) |961 |961 |0 |419 sec  (00:06:59) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  729 sec (00:12:09) |  120 sec (00:02:00) |   23 sec (00:00:23) |   18 sec (00:00:18) | 1847 sec (00:30:47) |   19 sec (00:00:19) | 2765 sec (00:46:05) |

timothyklemm commented on esp/esdllib/esdl_script.cpp:
```diff
@@ -136,26 +141,133 @@ class CEsdlTransformOperationParameter : public CEsdlTransformOperationVariable
+    }
+};
+
+class CEsdlTransformOperationStoreValue : public CEsdlTransformOperationSetSectionAttributeBase
```
Would there be value allowing more complex values to be stored? If we think we might want to store more structured data in the future, we might not want to rely on attributes now. I'm not saying we will. I'm just asking if it was considered.

afishbeck replied:
yes I did think about it.  I'm just leaving the actual content of the section for future consideration.  If we need something more than key value pairs I'd like to understand it before adding support... but will do so if we need it.

Also it will be more valuable when we have better ways of building structures.  Perhaps after we've moved away from IPtree on the right hand side.

timothyklemm replied:
I guess I was thinking that more complex data would suggest child elements. We could use child elements with name and value for the current key value pairs, then for more complex data replace the value attribute with content. Just thought it might be more consistent when we grow into it. Either way works.

timothyklemm commented on esp/esdllib/esdl_xpath_extensions_libxml.cpp:
```diff
@@ -198,9 +367,19 @@ void registerEsdlXPathExtensions(IXpathContext *xpathContext, IEspContext *conte
+        xpathContext->registerFunction(nullptr, "storedValueExists", (void  *)storedValueExistsFunction);
+        xpathContext->registerFunction(nullptr, "getLogProfile", (void  *)getLogProfileFunction);
+        xpathContext->registerFunction(nullptr, "getLogOption", (void  *)getLogOptionFunction);
+        xpathContext->registerFunction(nullptr, "logOptionExists", (void  *)logOptionExistsFunction);
```
Because I know the implementation details I could use this to test for a selected log profile, but should I have to? It seems like logProfileExists should be defined.

afishbeck replied:
You could just check if getLogProfile returns content.  I'll think about it.

afishbeck replied:
Running out of time for the initial 7.10 RC.  We can add a new xpath function for this later.  Just get if getLogProfile has content for now.

timothyklemm commented on esp/esdllib/esdl_xpath_extensions_libxml.cpp:
```diff
@@ -198,9 +367,19 @@ void registerEsdlXPathExtensions(IXpathContext *xpathContext, IEspContext *conte
+        xpathContext->registerFunction(nullptr, "logOptionExists", (void  *)logOptionExistsFunction);
     }
 
     xpathContext->registerFunction("urn:hpcc:esdl:script", "validateFeaturesAccess", (void  *)validateFeaturesAccessFunction);
     xpathContext->registerFunction("urn:hpcc:esdl:script", "secureAccessFlags", (void  *)secureAccessFlagsFunction);
     xpathContext->registerFunction("urn:hpcc:esdl:script", "getFeatureSecAccessFlags", (void  *)getFeatureSecAccessFlagsFunction);
+    xpathContext->registerFunction("urn:hpcc:esdl:script", "getStoredStringValue", (void  *)getStoredStringValueFunction);
+    xpathContext->registerFunction("urn:hpcc:esdl:script", "storedValueExists", (void  *)storedValueExistsFunction);
+    xpathContext->registerFunction("urn:hpcc:esdl:script", "getLogProfile", (void  *)getLogProfileFunction);
+    xpathContext->registerFunction("urn:hpcc:esdl:script", "getLogOption", (void  *)getLogOptionFunction);
+    xpathContext->registerFunction("urn:hpcc:esdl:script", "logOptionExists", (void  *)logOptionExistsFunction);
```
This looks like a candidate for a helper function, passing the urn and registering all functions. As the function list grows it will be more helpful.

afishbeck replied:
agreed the list is getting long enough.

timothyklemm commented on esp/esdllib/esdl_script.cpp:
```diff
@@ -136,26 +141,133 @@ class CEsdlTransformOperationParameter : public CEsdlTransformOperationVariable
+    }
+
+    virtual ~CEsdlTransformOperationStoreValue(){}
+    const char *getSectionName() override {return "store";}
```
The string literals should be defined names. Define the convention now, and hopefully avoid inconsistencies with future extensions.

afishbeck replied:
Ok agreed.

timothyklemm commented on esp/esdllib/esdl_script.cpp:
```diff
@@ -226,6 +338,152 @@ class CEsdlTransformOperationSetValue : public CEsdlTransformOperationBase
+
+            //make sure we only remove one.  we can add a multi-node support in the future (when we are libxml2 based?)
+            //don't want users to depend on quirky behavior
+            path.append("[1]"); //don't need to check as multiple [] are ok
```
With the libxml2 version of this, my intent was to pass an attribute specifying if all occurrences were to be affected. If not applied to all, a multiple node nodeset would be considered a failure. Just picking one of many seems like quirky behavior best to be avoided.

afishbeck replied:
I can throw an exception if multiple... and leave it to the script writer to add the [1] if desired.  There are cases where since this state can creep up externally that causing the whole transaction to fail could be undesirable.  This will mean script writers need to be diligent of this at all times.

timothyklemm commented on esp/esdllib/esdl_script.cpp:
```diff
@@ -226,6 +338,152 @@ class CEsdlTransformOperationSetValue : public CEsdlTransformOperationBase
+            else
+                path.set(m_target);
+
+            if (isWildString(path))
```
It seems like we should just evaluate the path and delete a single node, if that is what getElements gives us. Unless this required because getElements can't evaluate the wildcard.

afishbeck replied:
We are not calling getElements. If I remember correclty removeProp did not like wildcards.

timothyklemm replied:
I know it isn't using getElements. This was my way of suggesting it should use getElements. I just think taking '*' away, which is useful for more than looking for multiples, is undesirable.

timothyklemm commented on esp/esdllib/esdl_script.cpp:
```diff
@@ -226,6 +338,152 @@ class CEsdlTransformOperationSetValue : public CEsdlTransformOperationBase
+
+            //make sure we only remove one.  we can add a multi-node support in the future (when we are libxml2 based?)
+            //don't want users to depend on quirky behavior
+            path.append("[1]"); //don't need to check as multiple [] are ok
```
Same comment as with rename-node. Also, can this be used within a for-each to remove all matching nodes?

timothyklemm commented on esp/esdllib/esdl_script.cpp:
```diff
@@ -769,155 +1046,145 @@ class CEsdlCustomTransform : public CInterfaceOf<IEsdlCustomTransform>
-        level = (unsigned) tgtcfg->getPropInt("@traceLevel", level);
+    if (isEmptyString(srcSection)||isEmptyString(tgtSection))
+        return;
+    level = scriptCtx->getXPathInt64("target/*/@traceLevel", level);
```
LogLevel is unsigned, not __int64. This should be generating a compiler warning for potential loss of data.

afishbeck replied:
how do we get away with less than 9,223,372,036,854,775,807 log levels?

timothyklemm replied:
I didn't think implementing 128 bit integers was within the scope of this story, but if you think more options make sense...

afishbeck replied:
So the getXPathInt64 is so we can support other values that might require large ints.  But I'll add a specific cast here so the compiler wont care.  I don't want to add mulitple getInt functions if I don't have to.

timothyklemm commented on esp/esdllib/esdl_script.cpp:
```diff
@@ -769,155 +1046,145 @@ class CEsdlCustomTransform : public CInterfaceOf<IEsdlCustomTransform>
-            DBGLOG(1,"MODIFIED content: %s", content.str());
+    if (level >= LogMax)
+    {
+        scriptCtx->toXML(content.clear(), tgtSection);
```
This is helpful, but I found it to be more helpful changing this to the entire document instead of one section. With updates to sections outside of this one possible, seeing the whole document may be more useful.

afishbeck replied:
I was emulating what was already logged here, but I can make it the whole blob for now.  That may end up being to verbose, especially as it grows, but we can change it in the future.

timothyklemm commented on system/xmllib/libxml_xpathprocessor.cpp:
```diff
@@ -775,6 +789,352 @@ extern ICompiledXpath* compileXpath(const char * xpath)
     return new CLibCompiledXpath(xpath);
 }
 
+xmlNsPtr createNsFromXmlns(xmlNodePtr node, const char *attname, const char *uri, xmlNsPtr ns)
```
Am I missing something, or is this unused?

afishbeck replied:
It's probably from a prior iteration, I'll check.

timothyklemm commented on system/xmllib/libxml_xpathprocessor.cpp:
```diff
@@ -775,6 +789,352 @@ extern ICompiledXpath* compileXpath(const char * xpath)
+    else
+    {
+        xmlNsPtr ns = xmlSearchNs(node->doc, node, (const xmlChar *)prefix.str());
+        if (!ns) //invalid but let's repair it
+            ns = xmlNewNs(node, (const xmlChar *)"urn:hpcc:unknown", (const xmlChar *)prefix.str());
```
What use case are we enabling that cannot be described as "sloppy"? There may be a case we need to handle this way, but I can't think of it.

afishbeck replied:
I don't want to have to revisit just because we've got a ptree from somewhere that we want to be able to process and ptrees are notoriously bad about namespacing.  This is so I never have to look at this code again. ;)

timothyklemm commented on system/xmllib/xpathprocessor.hpp:
```diff
@@ -81,4 +81,22 @@ class CXpathContextScope : CInterface
+    virtual void setAttribute(const char *section, const char *name, const char *value) = 0;
+    virtual const char *queryAttribute(const char *section, const char *name) = 0;
+    virtual const char *getAttribute(const char *section, const char *name, StringBuffer &s) = 0;
+    virtual const char *getXPathString(const char *xpath, StringBuffer &s) const = 0;
+    virtual __int64 getXPathInt64(const char *xpath, __int64 dft=0) const = 0;
+    virtual bool getXPathBool(const char *xpath, bool dft=false) const = 0;
```
I know this naming fits existing code style (like PTree), but I'm not a fan. My preference would be to follow the design of IEspSecureContext. It would handle unsigned versus signed, or floating point versus integer, or short versus long.

A similar change could be made to getAttribute and setAttribute.

afishbeck replied:
Can consider refactoring in the future.  At least this is following a precedent.  I don't think this is currently an interface that we can't easily change.

richardkchapman commented:
@afishbeck I think this is with you at the moment?

afishbeckln committed:
HPCC-24218 Maintain cross ESDL transform state and add new functions

Use a scriptContext to support keeping script state for the entire transaction.

Allowing for the upcomming addition of multiple script entry points.

Add new script functions to add to scriptContext state:
    set-log-profile - signal which logging profile should be followed
    set-log-option - control logging options
    store-value - store values that can be accessed by other scripts

Add new script functions (single node operations for now):
   remove-node - remove a single node from the target tree
   rename-node - rename a single node on the target tree

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexisrisk.com>

# Conflicts:
#	esp/esdllib/esdl_script.cpp
#	system/xmllib/libxml_xpathprocessor.cpp
#	testing/unittests/esdltests.cpp

afishbeck commented:
@asselitx @timothyklemm updated, please review again.  Trying to get this in tomorrows release candidate.

HPCCSmoketest commented:
Process of PR-13848, label: esdlScriptContextData starts now.
The reason of this test is: Forced to re-test.
Commit ID: e24e6ff1850d7cae46a5a6eab83a50bc42088e72
Estimated completion time is ~0.8 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-80.ca-central-1.compute.internal

timothyklemm commented on esp/esdllib/esdl_xpath_extensions_libxml.cpp:
```diff
@@ -176,7 +189,175 @@ static void getFeatureSecAccessFlagsFunction (xmlXPathParserContextPtr ctxt, int
+    if (xmlXPathCheckError(ctxt)) //includes null check
+        return;
+
+    const char *value = scriptContext->queryAttribute("logging", (const char *)namestring);
```
This comment applies to multiple functions. The literal, or literals elsewhere, should be using the defined names.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-80.ca-central-1.compute.internal
Sha: e24e6ff1850d7cae46a5a6eab83a50bc42088e72
Build: success
Milestone:Install hpccsystems-platform-community_7.9.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |144 |144 |0 |0 |0 |72 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |27 sec  (00:00:27) |
|setup (thor) |11 |11 |0 |42 sec  (00:00:42) |
|setup (roxie) |11 |11 |0 |23 sec  (00:00:23) |
|test (hthor) |883 |883 |0 |307 sec  (00:05:07) |
|test (thor) |821 |821 |0 |762 sec  (00:12:42) |
|test (roxie) |962 |962 |0 |405 sec  (00:06:45) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  723 sec (00:12:03) |  119 sec (00:01:59) |   23 sec (00:00:23) |   19 sec (00:00:19) | 1823 sec (00:30:23) |   18 sec (00:00:18) | 2734 sec (00:45:34) |

timothyklemm commented on system/xmllib/libxml_xpathprocessor.cpp:
```diff
@@ -833,6 +848,346 @@ extern ICompiledXpath* compileXpath(const char * xpath)
+            return sect;
+        return xmlNewChild(root, nullptr, (const xmlChar *) name, nullptr);
+    }
+    void removeSection(const char *name)
```
With this name I would I would expect it to call getSection and delete the node. The implementation looks like it removes any arbitrary nodes, which may include a section node.

afishbeck replied:
This is fault tolerant if for some reason multiple nodes exist.  Should never happen but might as well clean them up.

timothyklemm commented on system/xmllib/libxml_xpathprocessor.cpp:
```diff
@@ -833,6 +848,346 @@ extern ICompiledXpath* compileXpath(const char * xpath)
+        {
+            xmlNodeSetPtr ns = eval->nodesetval;
+            int count = xmlXPathNodeSetGetLength(ns);
+            for(int i=0; i<count; i++)
```
Based on a comment in the example at http://www.xmlsoft.org/examples/xpath2.c, should the loop traverse in the opposite order? The evaluation could return a node and that node's descendant, and freeing the first node frees the descendant, invalidating the node reference.

afishbeck replied:
Agreed.

timothyklemm commented on system/xmllib/libxml_xpathprocessor.cpp:
```diff
@@ -833,6 +848,346 @@ extern ICompiledXpath* compileXpath(const char * xpath)
+    }
+    xmlNodePtr replaceSection(const char *name)
+    {
+        removeSection(name);
+        return xmlNewChild(root, nullptr, (const xmlChar *) name, nullptr);
```
Should there be verification that 'name' is an element name and not a path?

timothyklemm replied:
If setContent is called with an XPath and not a section name, replaceSection will act on the XPath by allowing removeSection to remove all matching occurrences. The new child of 'root' might not be created, but we've already potentially destroyed the document.

Whether it's here or in setContent, it seems like we need to verify that we're manipulating a section and not an arbitrary location. Otherwise it seems like a loophole waiting to be abused.

afishbeck replied:
I hope that people who knowingly try to pass in an xpath for the name get what they deserve when the document is destroyed.  But I do now check name is not a path via the call to sanityCheckSectionName().

afishbeck commented:
@timothyklemm updated based on your comments.

HPCCSmoketest commented:
Process of PR-13848, label: esdlScriptContextData starts now.
The reason of this test is: Forced to re-test.
Commit ID: 01fec1c0edf7d7ffdc530fc0d9c88c1a7656142e
Estimated completion time is ~0.8 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-135.ca-central-1.compute.internal

timothyklemm commented on esp/esdllib/esdl_xpath_extensions_libxml.cpp:
```diff
@@ -176,7 +189,175 @@ static void getFeatureSecAccessFlagsFunction (xmlXPathParserContextPtr ctxt, int
+    if (xmlXPathCheckError(ctxt)) //includes null check
+        return;
+
+    const char *value = scriptContext->queryAttribute("store", (const char *)namestring);
```
"store" should be ESDLScriptCtxSection_Store.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-135.ca-central-1.compute.internal
Sha: 01fec1c0edf7d7ffdc530fc0d9c88c1a7656142e
Build: success
Milestone:Install hpccsystems-platform-community_7.9.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |144 |144 |0 |0 |0 |73 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |26 sec  (00:00:26) |
|setup (thor) |11 |11 |0 |42 sec  (00:00:42) |
|setup (roxie) |11 |11 |0 |20 sec  (00:00:20) |
|test (hthor) |883 |883 |0 |305 sec  (00:05:05) |
|test (thor) |821 |821 |0 |763 sec  (00:12:43) |
|test (roxie) |962 |962 |0 |382 sec  (00:06:22) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  728 sec (00:12:08) |  117 sec (00:01:57) |   23 sec (00:00:23) |   20 sec (00:00:20) | 1796 sec (00:29:56) |   18 sec (00:00:18) | 2711 sec (00:45:11) |

HPCCSmoketest commented:
Process of PR-13848, label: esdlScriptContextData starts now.
The reason of this test is: Forced to re-test.
Commit ID: 31e8227846bde68f1861f727b45c50519e63a996
Estimated completion time is ~0.8 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-141.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-141.ca-central-1.compute.internal
Sha: 31e8227846bde68f1861f727b45c50519e63a996
Build: success
Milestone:Install hpccsystems-platform-community_7.9.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |144 |144 |0 |0 |0 |74 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |27 sec  (00:00:27) |
|setup (thor) |11 |11 |0 |45 sec  (00:00:45) |
|setup (roxie) |11 |11 |0 |24 sec  (00:00:24) |
|test (hthor) |883 |883 |0 |306 sec  (00:05:06) |
|test (thor) |821 |821 |0 |766 sec  (00:12:46) |
|test (roxie) |962 |962 |0 |407 sec  (00:06:47) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  712 sec (00:11:52) |  127 sec (00:02:07) |   22 sec (00:00:22) |   20 sec (00:00:20) | 1832 sec (00:30:32) |   20 sec (00:00:20) | 2742 sec (00:45:42) |

asselitx commented:
@afishbeck Your changes look good and they pass the tests against existing transforms.

afishbeck commented:
@richardkchapman this can be merged and included in RC1.

HPCCSmoketest commented:
Process of PR-13848, label: esdlScriptContextData starts now.
The reason of this test is: New commit.
Commit ID: 31e8227846bde68f1861f727b45c50519e63a996
Estimated completion time is ~0.9 hour(s)
centos 7.8.2003 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-116.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.8.2003 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-116.ca-central-1.compute.internal
Sha: 31e8227846bde68f1861f727b45c50519e63a996
Build: success
Milestone:Install hpccsystems-platform-community_7.9.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |144 |144 |0 |0 |0 |76 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |9 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |28 sec  (00:00:28) |
|setup (thor) |11 |11 |0 |45 sec  (00:00:45) |
|setup (roxie) |11 |11 |0 |20 sec  (00:00:20) |
|test (hthor) |883 |883 |0 |589 sec  (00:09:49) |
|test (thor) |821 |821 |0 |1052 sec  (00:17:32) |
|test (roxie) |962 |962 |0 |723 sec  (00:12:03) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   14 sec (00:00:14) | 1026 sec (00:17:06) |  124 sec (00:02:04) |   23 sec (00:00:23) |   17 sec (00:00:17) | 2720 sec (00:45:20) |   18 sec (00:00:18) | 3942 sec (01:05:42) |
